sdk built-in screenshot feature (#422)
getsentry/develop: src/docs/sdk/features.mdx

useful because attachments could quickly eat up the users' disk space. Furthermore, Relay has a
[maximum size for attachments](https://docs.sentry.io/product/relay/options/), and we want to reduce unnecessary requests.

## Screenshots

When the user opts-in, if technically possible, take a screenshot of the application and include as an [attachments](#attachments) to the [envelope](/sdk/envelopes/) with the event.

> This is a best effort feature that only applies to SDKs with a user interface, such as Mobile and Desktop.

Author: ```suggestion
This feature only applies to SDKs with a user interface, such as Mobile and Desktop. 
In some environments such as native iOS, taking a screenshot requires the UI thread and in the event of a crash, that might not be available. So inherently this feature will be a best effort.
```

useful because attachments could quickly eat up the users' disk space. Furthermore, Relay has a
[maximum size for attachments](https://docs.sentry.io/product/relay/options/), and we want to reduce unnecessary requests.

## Screenshots

When the user opts-in, if technically possible, take a screenshot of the application and include as an [attachments](#attachments) to the [envelope](/sdk/envelopes/) with the event.

This feature only applies to SDKs with a user interface, such as Mobile and Desktop. 
In some environments such as native iOS, taking a screenshot requires the UI thread and in the event of a crash, that might not be available. So inherently this feature will be a best effort.
It's advised to provide this feature through a single option called `attachScreenshot`. That's a preferred way but in platforms such as Flutter, a wrapping widget is required so documentation can point users to that instead of the suggested option name.

The feature is achieved by adding an attachment with:

* File name `screenshot.jpg` or `screenshot.png`

> * Image size, if possible should stay below 2 MB but quality/size could be configurable

Author: ```suggestion
* Image size, if possible should stay below 2 MB but quality/size could be configurable
* `ContentType: image/jpg` or `ContentType: image/png`
```

useful because attachments could quickly eat up the users' disk space. Furthermore, Relay has a
[maximum size for attachments](https://docs.sentry.io/product/relay/options/), and we want to reduce unnecessary requests.

## Screenshots

> When the user opts-in, if technically possible, take a screenshot of the application and include as an [attachments](#attachments) to the [envelope](/sdk/envelopes/) with the event.

Reviewer: ```suggestion
When the user opts-in, if technically possible, take a screenshot of the application and include as an [attachment](#attachments) to the [envelope](/sdk/envelopes/) with the event.
```

> When the user opts-in, if technically possible, take a screenshot of the application and include as an [attachments](#attachments) to the [envelope](/sdk/envelopes/) with the event.

Reviewer 2: ```suggestion
When the user opts-in, if technically possible, take a screenshot of the application and include it as an [attachments](#attachments) to the [envelope](/sdk/envelopes/) with the event.
```

Author: Make sure to add a note on PII.

This feature only applies to SDKs with a user interface, such as Mobile and Desktop. 

> In some environments such as native iOS, taking a screenshot requires the UI thread and in the event of a crash, that might not be available. So inherently this feature will be a best effort.

Reviewer 2: ```suggestion
In some environments such as native iOS, taking a screenshot requires the UI thread and in the event of a crash, that might not be available. So inherently this feature will be a best effort solution.
```

> It's advised to provide this feature through a single option called `attachScreenshot`. That's a preferred way but in platforms such as Flutter, a wrapping widget is required so documentation can point users to that instead of the suggested option name.

Reviewer: ```suggestion
It's advised to provide this feature through a single option called `attachScreenshot`. That's the preferred way but in platforms such as Flutter, a wrapping widget is required so documentation can point users to that instead of the suggested option name.
```

> 

Reviewer 3: ```suggestion
Furthermore, if  `attachScreenshot` is enabled, the SDK should attach screenshots to events with levels higher or equal to error, to avoid attaching screenshots to debug/info events (the same could be configurable).

```
My suggestion.

Reviewer 4: instead of checking the event level, I'd rather consider if its an errored or crashed event via exception list/mechanism, that's how it works for session, maybe we do the same? https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/SentryEvent.java#L223-L249

Author: Similar to `attachStacktrace` I was hoping we'd attach the screenshot on all events.
Perhaps we should consider a `shouldAttachScreenshot` callback that receives the event as an argument?
If the callback is `null` we always attach, but it gives the user a chance to configure things like: _only attach on unhandled exception_ or other use cases

Reviewer 4: I guess we have to give the user a chance to not add the screenshot (either via callback or only for more important events) since we also detect events automatically and taking SS is visible by the user (the screen kinda flashes out for a few millis)

Reviewer: Seems reasonable to make it very clear and intentional when screenshots (in the order of megabytes) go out of the SDK to avoid excessive data consumption.

Author: Agree with this. Lets consider this a second iteration.

The feature is achieved by adding an attachment with:

> * File name `screenshot.jpg` or `screenshot.png`

Reviewer 3: should we mention that it should overwrite any existing attachment with that given name during the process of attaching the screenshot?

Reviewer 4: its a good point, or doing a sort of send it once and drop it, so the same attachment won't be attached to every following event

Author: I think we should just add one and if there's always one added, the front end will just pick one of them. It's an edge case I rather the SDK doesn't have custom code to look for stuff in the attachments list (it can also be added after our check though so it would anwyay be best effort)

Reviewer 4: so if there were 3 events before, the 4th event contains the 3 previous attachments + the new one? I'd say it's burning users' data and not that nice.

Author: The SDK should add a single attachment when capturing the event. It'll not keep adding to the scope. Each capture includes a fresh screenshot. If the user had already added one manually with the same name, that's just an edge case I don't think worth spending too many cycles on.

* Image size, if possible should stay below 2 MB but quality/size could be configurable

> * `ContentType: image/jpg` or `ContentType: image/png`

Author: ```suggestion
* `ContentType: image/jpg` or `ContentType: image/png`

Whenever possible, avoid adding the attachment altogether if taking the screenshot fails. Alternatively, when streaming, it's possible the envelope header was already flushed through before the attempt to take the screenshot happens. In this case, a 0 byte attachment will be included. In that case, Sentry will not show a screenshot preview.
```

useful because attachments could quickly eat up the users' disk space. Furthermore, Relay has a
[maximum size for attachments](https://docs.sentry.io/product/relay/options/), and we want to reduce unnecessary requests.

## Screenshots

When the user opts-in, if technically possible, take a screenshot of the application and include as an [attachment](#attachments) to the [envelope](/sdk/envelopes/) with the event.

This feature only applies to SDKs with a user interface, such as Mobile and Desktop. 
In some environments such as native iOS, taking a screenshot requires the UI thread and in the event of a crash, that might not be available. So inherently this feature will be a best effort solution.
It's advised to provide this feature through a single option called `attachScreenshot`. That's the preferred way but in platforms such as Flutter, a wrapping widget is required so documentation can point users to that instead of the suggested option name.

The feature is achieved by adding an attachment with:

* File name `screenshot.jpg` or `screenshot.png`
* Image size, if possible should stay below 2 MB but quality/size could be configurable

> * `ContentType: image/jpg` or `ContentType: image/png`

Reviewer 2: Any particular reason why we need the `ContentType`? Does relay/the sentry product treat this differently>

Author: At the end we didn't rely on it at all for the preview. It's just file extension and file name. ContentType was @mitsuhiko's idea so maybe he can help us with a direction here

Reviewer 5: If no content type is set then sentry will try to guess it. I would generally prefer we send the content type explicitly.
